pkg/spanconfig/spanconfigsqlwatcher/sql_watcher_test.go

```diff
@@ -0,0 +1,171 @@
+// changes made to system.descriptor or system.zones.
+func TestSQLWatcher(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	type expectedUpdate struct {
+		expectedSpan    roachpb.Span
+		expectedDeleted bool
+	}
+
+	testCases := []struct {
+		setup           string
+		stmt            string
+		expectedUpdates []expectedUpdate
+	}{
+		{
+			stmt: "CREATE TABLE t()",
+			expectedUpdates: []expectedUpdate{
+				{
+					expectedSpan: roachpb.Span{
+						Key:    keys.SystemSQLCodec.TablePrefix(52),
+						EndKey: keys.SystemSQLCodec.TablePrefix(52).PrefixEnd(),
+					},
+					expectedDeleted: false,
+				},
+			},
+		},
+		{
+			setup: "CREATE table t2()",
+			stmt:  "ALTER TABLE t2 CONFIGURE ZONE USING num_replicas=5",
+			expectedUpdates: []expectedUpdate{
+				{
+					expectedSpan: roachpb.Span{
+						Key:    keys.SystemSQLCodec.TablePrefix(53),
+						EndKey: keys.SystemSQLCodec.TablePrefix(53).PrefixEnd(),
+					},
+					expectedDeleted: false,
+				},
+			},
+		},
+		{
+			setup: "CREATE TABLE t3(); ALTER TABLE t3 CONFIGURE ZONE USING gc.ttlseconds = 1;",
+			stmt:  "DROP TABLE t3",
+			expectedUpdates: []expectedUpdate{
+				{
+					expectedSpan: roachpb.Span{
+						Key:    keys.SystemSQLCodec.TablePrefix(54),
+						EndKey: keys.SystemSQLCodec.TablePrefix(54).PrefixEnd(),
+					},
+					expectedDeleted: false,
+				},
+				{
+					expectedSpan: roachpb.Span{
+						Key:    keys.SystemSQLCodec.TablePrefix(54),
+						EndKey: keys.SystemSQLCodec.TablePrefix(54).PrefixEnd(),
+					},
+					expectedDeleted: true,
+				},
+			},
+		},
+		{
+			setup: "CREATE DATABASE d; CREATE TABLE d.t1(); CREATE TABLE d.t2();",
+			stmt:  "ALTER DATABASE d CONFIGURE ZONE USING num_replicas=5",
+			expectedUpdates: []expectedUpdate{
+				{
+					expectedSpan: roachpb.Span{
+						Key:    keys.SystemSQLCodec.TablePrefix(56),
+						EndKey: keys.SystemSQLCodec.TablePrefix(56).PrefixEnd(),
+					},
+					expectedDeleted: false,
+				},
+				{
+					expectedSpan: roachpb.Span{
+						Key:    keys.SystemSQLCodec.TablePrefix(57),
+						EndKey: keys.SystemSQLCodec.TablePrefix(57).PrefixEnd(),
+					},
+					expectedDeleted: false,
+				},
+			},
+		},
+	}
+
+	ctx := context.Background()
+	tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{
+			Knobs: base.TestingKnobs{
+				SpanConfig: &spanconfig.TestingKnobs{
+					ManagerDisableJobCreation: true, // disable the automatic job creation
+				},
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+			},
+		},
+	})
+	defer tc.Stopper().Stop(ctx)
+
+	ts := tc.Server(0 /* idx */)
+	sqlWatcher := spanconfigsqlwatcher.New(
+		keys.SystemSQLCodec,
+		ts.DB(),
+		ts.ClusterSettings(),
+		ts.RangeFeedFactory().(*rangefeed.Factory),
+		ts.Clock(),
+		ts.InternalExecutor().(*sql.InternalExecutor),
+		ts.LeaseManager().(*lease.Manager),
+		tc.Stopper(),
+		&spanconfig.TestingKnobs{
+			SQLWatcherDisableInitialScan: true,
```

Comment:
Should have a variant with the initial as well scan, just to see it's what we expect.